### PR TITLE
Add Metal support for Qwen3-4B generation

### PR DESCRIPTION
**Pull request by spinlocked**
Extend MetalRuntime with the runtime APIs needed for loading safetensors, managing persistent KV-cache buffers, round-tripping output buffers back into inputs, and reading logits during autoregressive decoding.

Update the Qwen example to support both CUDA and Metal through mutually exclusive CUDA and Metal feature flags.

### CODE REVIEW

**Comment by jafioti** [comment]
@spinlocked  This is great work! Is it possible instead of having two seperate examples with a third common crate, to just have one example with the cuda and metal stuff gated with mutually exclusive feature flags, like "metal" and "cuda"?

**Comment by spinlocked** [comment]
@jafioti Yes, that does seem better. I’ve updated the changes to do that.

**Comment by jafioti** [comment]
@spinlocked did you want to add more to this PR or want me to review it now?

**Comment by spinlocked** [comment]
@jafioti Should be ready for review

**Comment by jafioti** [comment]
looks great!
